### DISCUSSION TITLE
Implementing a Particle Shader Effect

### DISCUSSION THREAD

**Discussion by ixxie** [pmndrs/postprocessing#477] [Q&A]
I'm quite new to postprocessing as well as lower-level WebGL stuff, so forgive the lack of nuance / precision in my question.

I'm interested in implementing a particle morph effect; I found two relevant examples using custom shaders:

  - **Colorver.se** - [demo](https://colorver.se) // [article (google trans.)](https://studio--jt-co-kr.translate.goog/projects/%EC%BB%AC%EB%9F%AC%EB%B2%84%EC%8A%A4/?_x_tr_sl=auto&_x_tr_tl=en&_x_tr_hl=en&_x_tr_pto=wapp)
  - **Interactive Particles** - [demo](https://tympanus.net/Tutorials/InteractiveParticles/) // [article](https://tympanus.net/codrops/2019/01/17/interactive-particles-with-three-js/) // [repo](https://github.com/brunoimbrizi/interactive-particles)
  
  I'm asking for advice on how such an effect could be implemented in postprocessing. I can try and port the interactive particles shader for example, but I am not sure where to put it (a `ShaderPass`?).
  
  Even better would be an example of how this is done with postprocessing, or any other relevant examples, but I know this may be a hard-ask.

**Comment by vanruesc** [comment]
I'd suggest creating a new Pass for this. `ShaderPass` is better suited for quick testing or very simple fullscreen shader effects. 

Then you could use two or three render targets to calculate the particle positions - two buffers for read/write (emulating [transform feedback](https://github.com/mrdoob/three.js/pull/14503)) and one buffer to copy the current result to so that it can be read by the particle vertex shader. This would keep all calculations and data on the GPU.

A simple GPGPU pass would just extend `Pass`, set up a few buffers and render a fullscreen material:

```js
export class ParticlePositionPass extends Pass {

	constructor() {

		super("ParticlePass");

		this.fullscreenMaterial = new ParticlePositionMaterial();
		this.needsSwap = false;

		this.renderTarget = new WebGLRenderTarget(X, Y, {
			type: HalfFloatType,
			minFilter: NearestFilter,
			magFilter: NearestFilter,
			depthBuffer: false
		});

		...

	}

	get positionTexture() {

		return this.renderTarget.texture;

	}

	render(renderer, inputBuffer, outputBuffer, deltaTime, stencilTest) {

		renderer.setRenderTarget(this.renderTarget);
		renderer.render(this.scene, this.camera);

		...

	}

}
```

**Comment by ixxie** [comment]
Thank you for this outline @vanruesc; I'm a little bit out of my depth here since I'm new to this lower-level WebGL stuff. I'm struggling with concepts like buffer and render target. 

Could you recommend some resources that would be suitable for me to grasp these basic concepts?
